simplefoc/Arduino-FOC
# STM32 HAL-only PWM driver

This is an initial attempt to port the PWM driver for STM32 to HAL only. Until now, our PWM driver has used the stm32duino HardwareTimer implementation. This PR replaces all references to the HardwareTimer API with direct calls to HAL or STM32 LL APIs.

Why?

- HardwareTimer is not supported on the Arduino H7 boards (Portenta, Nicla, Giga) which use Mbed. Therefore our standard STM32 driver was not being used there. Instead we had the "portenta" driver specifically for the portenta, but it was not far in development. It's a shame that we were not supporting the most powerful Arduino boards properly. With this HAL-only driver it compiles on Arduino Portenta as well, and supports all the features of the STM32 driver.
- HardwareTimer was restrictive in that it has an opinionated API which did not always suit our purposes. So we were using HAL and LL calls in several places anyway.
- HardwareTimer is less performant - it performs lots of checks and maintains internal state we don't need.

Additionally, the HAL only driver implements a number of other improvements over the HardwareTimer version:

- better multi-motor support
- improved PWM period setting - faster PWM periods possible on fast hardware and maximum possible range used for desired frequency
- better performance when setting PWM values: fewer floating point operations and comparisons
- aligning timers is now per motor rather than globally
- when using a single timer only in any PWM mode, the timer update is inhibited while the PWM values are being set. After all values are set, PWM updates are reactivated. This means PWM updates always occur in sync for all phases when a single timer is used.
- when using software 6-PWM also on multiple timers, the timer updates are inhibited while the PWM values are being set. This guarantees glitch-free software 6-PWM.
- aligned timers are configured in gated mode, meaning slaves are automatically started and stopped with the master timer. This guarantees timers will stay aligned

This PR is still in draft. Open issues to address:

- test on Portenta, Giga and other H7 platforms
- fix current sensing code for L4, F7, F1 and other series that I did not yet fix
- test in combination with current sensing
- see if some of the MBED only code can be replaced by code already included (discovering timer frequencies)
- add attributions for the code adapted from Arduino HardwareTimer
